### DISCUSSION TITLE
TableBrowser with header

### DISCUSSION THREAD

**Discussion by pamoroso** [Interlisp/medley#1808] [General]
To learn how to use TableBrowser I wrote this simple Interlisp program: [TBDEMO.TXT](https://github.com/user-attachments/files/16651530/TBDEMO.TXT)

Evaluating `(TBD.INIT)` creates a table with 3 columns and populates it with sample items. The table is supposed to have this header:

```
COLUMN 1           COLUMN 2            COLUMN 3
```

However, the header area is completely blank and looks like this:

![tablebrowser](https://github.com/user-attachments/assets/ce17f2be-be24-4425-98fa-0c8c6c4b9a2a)

I did expect the header text to be partially offset or cropped but hoped its location would hint at how to fix it. However, I didn't expect the text wouldn't show up at all. Is there anything obviously wrong with my code?

I checked the File Browser code but the implementation of the header is a bit involved and I haven't figured how it works.

**Comment by nbriggs** [comment]
You've passed the atom 'HEADER to the table browser create function, instead of passing the window that is the value of the atom HEADER.

You've also put the header repaintfn on the main window instead of the heading window.

**Comment by pamoroso** [comment]
I changed `TBD.INIT` to evaluate `HEADER` and put the header repaint function on the header window instead of the main window:

```
(TBD.INIT
  [LAMBDA NIL                                                (* ; "Edited 19-Aug-2024 04:54 by PA")
                                                             (* ; "Edited  7-May-2024 10:45 by PA")
                                                             (* ; "Edited 22-Oct-2023 11:20 by PA")
    (LET* [(WINDOW (CREATEW NIL "Table Browser"))
           [HEADER (CREATEW '(50 50 150 20]
           (BROWSER (TB.MAKE.BROWSER NIL WINDOW `(PRINTFN TBD.PRINTFN TBHEADINGWINDOW ,HEADER]
          (WINDOWPROP HEADER 'REPAINTFN (FUNCTION TBD.REPAINT.HEADER))
          (ATTACHWINDOW HEADER WINDOW 'TOP 'JUSTIFY)
          [for I from 1 to 10 do (TB.INSERT.ITEM BROWSER (TBD.CREATE.RECORD (CONCAT "Column 1 " I)
                                                                (CONCAT "Column 2 " I)
                                                                (CONCAT "Column 3 " I]
          BROWSER])
```

The full code is here: [TBDEMO.TXT](https://github.com/user-attachments/files/16660051/TBDEMO.TXT)


Now the header does show up but slightly offset and partially erased. And the header doesn't repaint or scroll in sync with the main window, but it's something I can probably figure.

![tb-header](https://github.com/user-attachments/assets/081481a0-d0a2-49ff-93f9-856c1c05f9f6)

**Comment by pamoroso** [comment]
I rewrote the repaint and reshape functions of the header window like this:  [TBDEMO.TXT](https://github.com/user-attachments/files/21101112/TBDEMO.TXT)

```
(TBD.REPAINT.HEADER
  [LAMBDA (WINDOW REGION)
    (DSPRESET WINDOW)
    (PRINTOUT WINDOW " COLUMN 1" 21 "COLUMN 2" 41 "COLUMN 3"])

(TBD.RESHAPE.HEADER
  [LAMBDA (WINDOW REGION)
    (TBD.REPAINT.HEADER WINDOW]))
```

Now the header of the table browser displays the way I intended:

![tb-header1](https://github.com/user-attachments/assets/aff70c28-d3c9-4ae5-af33-a61a3dace99c)

However, the table browser and the header don't scroll in sync but each scrolls independently:

![tb-header2](https://github.com/user-attachments/assets/8d8c6f4d-d4aa-4e64-8410-5f0c74e38cda)

Reshaping has layout issues too:

![tb-header3](https://github.com/user-attachments/assets/3201477e-521e-419e-ac02-5cc7bda6f336)

Part or all of the issues may be fixed by properly setting the extent of the header window, which the code doesn't yet.
